### ISSUE TITLE
Various compilation issues

### ISSUE THREAD

**Issue by viciious** [mikke89/RmlUi#47]
Oddly enough but I seem to be unable to compile RmlUi in Travis for my own project and it seems rightfully so:

```
In file included from /home/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/../../Include/RmlUi/Core/Types.h:51:

/home/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/../../Include/RmlUi/Core/Containers/robin_hood.h:657:22: error: implicit instantiation of undefined template 'std::hash<Rml::Core::PropertyId>'

struct hash : public std::hash<T> {

                     ^

/home/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/../../Include/RmlUi/Core/Containers/robin_hood.h:743:11: note: in instantiation of template class 'robin_hood::hash<Rml::Core::PropertyId>' requested here

        : public Hash,

                 ^

/home/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/../../Include/RmlUi/Core/PropertyDictionary.h:87:14: note: in instantiation of template class 'robin_hood::detail::unordered_map<true, 80, Rml::Core::PropertyId, Rml::Core::Property, robin_hood::hash<Rml::Core::PropertyId>, std::equal_to<Rml::Core::PropertyId> >' requested here

        PropertyMap properties;

                    ^

/usr/lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/bits/functional_hash.h:58:12: note: template is declared here

    struct hash;
```

PropertyId is a enum class and using it as a key for unordered_map requires a specialized hash function but it appears RmlUi is providing one.

**Comment by mikke89** [comment]
Should work in C++14, but looks like a bug in older versions of gcc. Does the fix work for you?

**Comment by viciious** [comment]
Hold on, I'll report back in a few minutes.
Actually, the problem exhibited itself on clang7 so I'm not entirely sure that the old compiler is at fault here.

**Comment by viciious** [comment]
Yup, that worked. Thanks!

**Comment by viciious** [comment]
The same compiler produces a different error in a different place now though:

```
In file included from /home/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/PropertySpecification.cpp:35:

/home/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/IdNameMap.h:62:3: error: unknown type name 'ptrdiff_t'; did you mean '__gnu_cxx::ptrdiff_t'?

                ptrdiff_t cnt = std::count_if(name_map.begin(), name_map.end(), [](const String& name) { return !name.empty(); });

                ^~~~~~~~~

                __gnu_cxx::ptrdiff_t

/usr/lib/gcc/x86_64-linux-gnu/5.4.0/../../../../include/c++/5.4.0/ext/new_allocator.h:45:14: note: '__gnu_cxx::ptrdiff_t' declared here

  using std::ptrdiff_t;
```

Should I open a different issue for this?

**Comment by mikke89** [comment]
I'm glad it worked!

How about the latest commit? :)

**Comment by viciious** [comment]
That worked as well, thanks!

**Comment by viciious** [comment]
Sorry for lumping this here but I believe a separate issue would be excessive. MacOS clang seems happy about the following stuff:

```
/Users/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/../../Include/RmlUi/Core/TransformPrimitive.h:128:91: warning: suggest braces around initialization of subobject [-Wmissing-braces]

        TranslateX(float x, Property::Unit unit = Property::PX) noexcept : UnresolvedPrimitive({ NumericValue(x, unit) }) { }

                                                                                                 ^~~~~~~~~~~~~~~~~~~~~

                                                                                                 {                    }

/Users/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/../../Include/RmlUi/Core/TransformPrimitive.h:134:91: warning: suggest braces around initialization of subobject [-Wmissing-braces]

        TranslateY(float y, Property::Unit unit = Property::PX) noexcept : UnresolvedPrimitive({ NumericValue(y, unit) }) { }

                                                                                                 ^~~~~~~~~~~~~~~~~~~~~

                                                                                                 {                    }

/Users/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/../../Include/RmlUi/Core/TransformPrimitive.h:140:91: warning: suggest braces around initialization of subobject [-Wmissing-braces]

        TranslateZ(float z, Property::Unit unit = Property::PX) noexcept : UnresolvedPrimitive({ NumericValue(z, unit) }) { }

                                                                                                 ^~~~~~~~~~~~~~~~~~~~~

                                                                                                 {                    }

/Users/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/../../Include/RmlUi/Core/TransformPrimitive.h:146:102: warning: suggest braces around initialization of subobject [-Wmissing-braces]

        Translate2D(float x, float y, Property::Unit units = Property::PX) noexcept : UnresolvedPrimitive({ NumericValue(x, units), NumericValue(y, units) }) { }

                                                                                                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

                                                                                                            {                                             }

/Users/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/../../Include/RmlUi/Core/TransformPrimitive.h:152:95: warning: suggest braces around initialization of subobject [-Wmissing-braces]

        Translate3D(NumericValue x, NumericValue y, NumericValue z) noexcept : UnresolvedPrimitive({ x, y, z }) { }

                                                                                                     ^~~~~~~

                                                                                                     {      }

/Users/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/../../Include/RmlUi/Core/TransformPrimitive.h:153:111: warning: suggest braces around initialization of subobject [-Wmissing-braces]

        Translate3D(float x, float y, float z, Property::Unit units = Property::PX) noexcept : UnresolvedPrimitive({ NumericValue(x, units), NumericValue(y, units), NumericValue(z, units) }) { }

                                                                                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

                                                                                                                     {                                                                     }

/Users/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/../../Include/RmlUi/Core/TransformPrimitive.h:159:53: warning: suggest braces around initialization of subobject [-Wmissing-braces]

        ScaleX(float value) noexcept : ResolvedPrimitive({ value }) { }

                                                           ^~~~~

                                                           {    }

/Users/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/../../Include/RmlUi/Core/TransformPrimitive.h:165:53: warning: suggest braces around initialization of subobject [-Wmissing-braces]

        ScaleY(float value) noexcept : ResolvedPrimitive({ value }) { }

                                                           ^~~~~

                                                           {    }

/Users/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/../../Include/RmlUi/Core/TransformPrimitive.h:171:53: warning: suggest braces around initialization of subobject [-Wmissing-braces]

        ScaleZ(float value) noexcept : ResolvedPrimitive({ value }) { }

                                                           ^~~~~

                                                           {    }

/Users/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/../../Include/RmlUi/Core/TransformPrimitive.h:177:51: warning: suggest braces around initialization of subobject [-Wmissing-braces]

        Scale2D(float xy) noexcept : ResolvedPrimitive({ xy, xy }) { }

                                                         ^~~~~~

                                                         {     }

/Users/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/../../Include/RmlUi/Core/TransformPrimitive.h:178:59: warning: suggest braces around initialization of subobject [-Wmissing-braces]

        Scale2D(float x, float y) noexcept : ResolvedPrimitive({ x, y }) { }

                                                                 ^~~~

                                                                 {   }

/Users/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/../../Include/RmlUi/Core/TransformPrimitive.h:184:52: warning: suggest braces around initialization of subobject [-Wmissing-braces]

        Scale3D(float xyz) noexcept : ResolvedPrimitive({ xyz, xyz, xyz }) { }

                                                          ^~~~~~~~~~~~~

                                                          {            }

/Users/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/../../Include/RmlUi/Core/TransformPrimitive.h:185:68: warning: suggest braces around initialization of subobject [-Wmissing-braces]

        Scale3D(float x, float y, float z) noexcept : ResolvedPrimitive({ x, y, z }) { }

                                                                          ^~~~~~~

                                                                          {      }

/Users/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/../../Include/RmlUi/Core/TransformPrimitive.h:190:77: warning: suggest braces around initialization of subobject [-Wmissing-braces]

        RotateX(const NumericValue* values) noexcept : ResolvedPrimitive(values, { Property::RAD }) { }

                                                                                   ^~~~~~~~~~~~~

                                                                                   {            }

/Users/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/../../Include/RmlUi/Core/TransformPrimitive.h:191:91: warning: suggest braces around initialization of subobject [-Wmissing-braces]

        RotateX(float angle, Property::Unit unit = Property::DEG) noexcept : ResolvedPrimitive({ NumericValue{ angle, unit } }, { Property::RAD }) { }

                                                                                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~

                                                                                                 {                          }

/Users/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/../../Include/RmlUi/Core/TransformPrimitive.h:191:124: warning: suggest braces around initialization of subobject [-Wmissing-braces]

        RotateX(float angle, Property::Unit unit = Property::DEG) noexcept : ResolvedPrimitive({ NumericValue{ angle, unit } }, { Property::RAD }) { }

                                                                                                                                  ^~~~~~~~~~~~~

                                                                                                                                  {            }

/Users/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/../../Include/RmlUi/Core/TransformPrimitive.h:196:77: warning: suggest braces around initialization of subobject [-Wmissing-braces]

        RotateY(const NumericValue* values) noexcept : ResolvedPrimitive(values, { Property::RAD }) {}

                                                                                   ^~~~~~~~~~~~~

                                                                                   {            }

/Users/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/../../Include/RmlUi/Core/TransformPrimitive.h:197:91: warning: suggest braces around initialization of subobject [-Wmissing-braces]

        RotateY(float angle, Property::Unit unit = Property::DEG) noexcept : ResolvedPrimitive({ NumericValue{ angle, unit } }, { Property::RAD }) { }

                                                                                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~

                                                                                                 {                          }

/Users/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/../../Include/RmlUi/Core/TransformPrimitive.h:197:124: warning: suggest braces around initialization of subobject [-Wmissing-braces]

        RotateY(float angle, Property::Unit unit = Property::DEG) noexcept : ResolvedPrimitive({ NumericValue{ angle, unit } }, { Property::RAD }) { }

                                                                                                                                  ^~~~~~~~~~~~~

                                                                                                                                  {            }

/Users/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/../../Include/RmlUi/Core/TransformPrimitive.h:202:77: warning: suggest braces around initialization of subobject [-Wmissing-braces]

        RotateZ(const NumericValue* values) noexcept : ResolvedPrimitive(values, { Property::RAD }) { }

                                                                                   ^~~~~~~~~~~~~

                                                                                   {            }

/Users/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/../../Include/RmlUi/Core/TransformPrimitive.h:203:91: warning: suggest braces around initialization of subobject [-Wmissing-braces]

        RotateZ(float angle, Property::Unit unit = Property::DEG) noexcept : ResolvedPrimitive({ NumericValue{ angle, unit } }, { Property::RAD }) { }

                                                                                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~

                                                                                                 {                          }

/Users/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/../../Include/RmlUi/Core/TransformPrimitive.h:203:124: warning: suggest braces around initialization of subobject [-Wmissing-braces]

        RotateZ(float angle, Property::Unit unit = Property::DEG) noexcept : ResolvedPrimitive({ NumericValue{ angle, unit } }, { Property::RAD }) { }

                                                                                                                                  ^~~~~~~~~~~~~

                                                                                                                                  {            }

/Users/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/../../Include/RmlUi/Core/TransformPrimitive.h:208:78: warning: suggest braces around initialization of subobject [-Wmissing-braces]

        Rotate2D(const NumericValue* values) noexcept : ResolvedPrimitive(values, { Property::RAD }) { }

                                                                                    ^~~~~~~~~~~~~

                                                                                    {            }

/Users/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/../../Include/RmlUi/Core/TransformPrimitive.h:209:92: warning: suggest braces around initialization of subobject [-Wmissing-braces]

        Rotate2D(float angle, Property::Unit unit = Property::DEG) noexcept : ResolvedPrimitive({ NumericValue{ angle, unit } }, { Property::RAD }) { }

                                                                                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~

                                                                                                  {                          }

/Users/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/../../Include/RmlUi/Core/TransformPrimitive.h:209:125: warning: suggest braces around initialization of subobject [-Wmissing-braces]

        Rotate2D(float angle, Property::Unit unit = Property::DEG) noexcept : ResolvedPrimitive({ NumericValue{ angle, unit } }, { Property::RAD }) { }

                                                                                                                                   ^~~~~~~~~~~~~

                                                                                                                                   {            }

/Users/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/../../Include/RmlUi/Core/TransformPrimitive.h:214:78: warning: suggest braces around initialization of subobject [-Wmissing-braces]

        Rotate3D(const NumericValue* values) noexcept : ResolvedPrimitive(values, { Property::NUMBER, Property::NUMBER, Property::NUMBER, Property::RAD }) { }

                                                                                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

                                                                                    {                                                                  }

/Users/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/../../Include/RmlUi/Core/TransformPrimitive.h:216:25: warning: suggest braces around initialization of subobject [-Wmissing-braces]

                : ResolvedPrimitive({ NumericValue{x, Property::NUMBER}, NumericValue{y, Property::NUMBER}, NumericValue{z, Property::NUMBER}, NumericValue{angle, angle_unit} },

                                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

                                      {                                                                                                                                       }

/Users/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/../../Include/RmlUi/Core/TransformPrimitive.h:217:6: warning: suggest braces around initialization of subobject [-Wmissing-braces]

                        { Property::NUMBER, Property::NUMBER, Property::NUMBER, Property::RAD }) { }

                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

                          {                                                                  }

/Users/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/../../Include/RmlUi/Core/TransformPrimitive.h:222:75: warning: suggest braces around initialization of subobject [-Wmissing-braces]

        SkewX(const NumericValue* values) noexcept : ResolvedPrimitive(values, { Property::RAD }) { }

                                                                                 ^~~~~~~~~~~~~

                                                                                 {            }

/Users/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/../../Include/RmlUi/Core/TransformPrimitive.h:223:89: warning: suggest braces around initialization of subobject [-Wmissing-braces]

        SkewX(float angle, Property::Unit unit = Property::DEG) noexcept : ResolvedPrimitive({ NumericValue{ angle, unit } }, { Property::RAD }) { }

                                                                                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~

                                                                                               {                          }

/Users/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/../../Include/RmlUi/Core/TransformPrimitive.h:223:122: warning: suggest braces around initialization of subobject [-Wmissing-braces]

        SkewX(float angle, Property::Unit unit = Property::DEG) noexcept : ResolvedPrimitive({ NumericValue{ angle, unit } }, { Property::RAD }) { }

                                                                                                                                ^~~~~~~~~~~~~

                                                                                                                                {            }

/Users/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/../../Include/RmlUi/Core/TransformPrimitive.h:228:75: warning: suggest braces around initialization of subobject [-Wmissing-braces]

        SkewY(const NumericValue* values) noexcept : ResolvedPrimitive(values, { Property::RAD }) { }

                                                                                 ^~~~~~~~~~~~~

                                                                                 {            }

/Users/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/../../Include/RmlUi/Core/TransformPrimitive.h:229:89: warning: suggest braces around initialization of subobject [-Wmissing-braces]

        SkewY(float angle, Property::Unit unit = Property::DEG) noexcept : ResolvedPrimitive({ NumericValue{ angle, unit } }, { Property::RAD }) { }

                                                                                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~

                                                                                               {                          }

/Users/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/../../Include/RmlUi/Core/TransformPrimitive.h:229:122: warning: suggest braces around initialization of subobject [-Wmissing-braces]

        SkewY(float angle, Property::Unit unit = Property::DEG) noexcept : ResolvedPrimitive({ NumericValue{ angle, unit } }, { Property::RAD }) { }

                                                                                                                                ^~~~~~~~~~~~~

                                                                                                                                {            }

/Users/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/../../Include/RmlUi/Core/TransformPrimitive.h:234:76: warning: suggest braces around initialization of subobject [-Wmissing-braces]

        Skew2D(const NumericValue* values) noexcept : ResolvedPrimitive(values, { Property::RAD, Property::RAD }) { }

                                                                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~

                                                                                  {                           }

/Users/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/../../Include/RmlUi/Core/TransformPrimitive.h:235:94: warning: suggest braces around initialization of subobject [-Wmissing-braces]

        Skew2D(float x, float y, Property::Unit unit = Property::DEG) noexcept : ResolvedPrimitive({NumericValue{ x, unit }, { NumericValue{ y, unit }}}, {Property::RAD, Property::RAD}) { }

                                                                                                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

                                                                                                    {                                                  }

/Users/travis/build/Qfusion/qfusion/third-party/libRocket/Source/Core/../../Include/RmlUi/Core/TransformPrimitive.h:235:149: warning: suggest braces around initialization of subobject [-Wmissing-braces]

        Skew2D(float x, float y, Property::Unit unit = Property::DEG) noexcept : ResolvedPrimitive({NumericValue{ x, unit }, { NumericValue{ y, unit }}}, {Property::RAD, Property::RAD}) { }

                                                                                                                                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~

                                                                                                                                                           {                           }

37 warnings generated.
```

**Comment by viciious** [comment]
Please also consider adding the following includes to Types.h:

```
+#include <cstring>
+#include <algorithm>
```

These are implicitly included with third-party containers but if you compile RmlUI without them, depending on the include order, compilation will likely break.

**Comment by mikke89** [comment]
I think the missing braces warning are the same false warnings reported a while back (compiler bug, fixed in later versions of clang).

I agree with adding the includes, I see you already did it in the pull request so I will leave it alone :)

**Comment by viciious** [comment]
I'll make a separate PR for that :)
The commit you're seeing is a just a temporary bandaid

**Comment by mikke89** [comment]
Includes fixed in #48.

Can we close this for now, reopen perhaps if new ones emerge?

**Comment by viciious** [comment]
The original issue is fixed in #49 for Travis builds too

**Comment by mikke89** [comment]
Great!
